lib/datadog/appsec/waf.rb

```diff
@@ -501,24 +484,19 @@ def initialize(rule, limits: {}, obfuscator: {})
 
           @config = config_obj
 
-          ruleset_info = LibDDWAF::RuleSetInfo.new
+          ruleset_info_obj = Datadog::AppSec::WAF::LibDDWAF::Object.new
 
-          @handle_obj = Datadog::AppSec::WAF::LibDDWAF.ddwaf_init(rule_obj, config_obj, ruleset_info)
+          @handle_obj = Datadog::AppSec::WAF::LibDDWAF.ddwaf_init(rule_obj, config_obj, ruleset_info_obj)
 
-          @ruleset_info = {
-            loaded: ruleset_info[:loaded],
-            failed: ruleset_info[:failed],
-            errors: WAF.object_to_ruby(ruleset_info[:errors]),
-            version: ruleset_info[:version],
-          }
+          @ruleset_info = Datadog::AppSec::WAF.object_to_ruby(ruleset_info_obj)
 
           if @handle_obj.null?
             fail LibDDWAF::Error.new('Could not create handle', ruleset_info: @ruleset_info)
           end
 
           validate!
         ensure
-          Datadog::AppSec::WAF::LibDDWAF.ddwaf_ruleset_info_free(ruleset_info) if ruleset_info
+          Datadog::AppSec::WAF::LibDDWAF.ddwaf_object_free(ruleset_info_obj) if ruleset_info_obj
```

Comment:
I'm curious about this one -- according to the release notes, it says that the new object (emphasis mine):

> This object is now provided as a parameter to both ddwaf_init and ddwaf_update, **and it should be allocated by the caller (e.g. stack-allocated as a local variable)**

We're not doing stack allocation (I see we're creating it on the Ruby heap), but won't Ruby correctly manage this object? E.g. do we still need to use `ddwaf_object_free` for an object that Ruby itself is managing?

(I'm lacking a bunch of context on how we manage these objects, so it may be fine, but I'm curious :P)

Reply:
I believe we are indeed allocating on the stack.

The Object class is a FFI struct:
```
class Object < ::FFI::Struct
    layout :parameterName,       :charptr,
           :parameterNameLength, :uint64,
           :valueUnion,          ObjectValueUnion,
           :nbEntries,           :uint64,
           :type,                :ddwaf_obj_type
  end
```

From the [FFI wiki](https://github.com/ffi/ffi/wiki/Structs): 

```ruby
class SimpleStruct < FFI::Struct
  layout :value, :double
end

a = SimpleStruct.new # does a malloc on one of those structs

a[:value] = 32 # sets its internal value to 32
```

After creating the handle we convert the FFI struct into a ruby object that we store in the Handle instance and make sure to free the memory of the allocated object using `ddwaf_object_free `

Reply:
Are you sure about the stack allocation stuff? We're creating an actual Ruby object with `Datadog::AppSec::WAF::LibDDWAF::Object.new`, and I'm pretty sure Ruby objects can't be stack allocated :P

(I actually think we're heap allocating in Ruby, but libddwaf doesn't mind, and the Ruby garbage collector is the one taking care of cleaning that part)

As for the rest, I think intuitively it kinda makes some sense although I'll admit I'm mentally hand-waving some of it haha, I don't think I fully understand how exactly each part of the memory is getting managed.